wasm/derive/src/lib.rs

```diff
@@ -0,0 +1,70 @@
+//! Exposes macros which facilitate writing smartcontracts
+
+#![allow(clippy::str_to_string)]
+
+use proc_macro::TokenStream;
+use proc_macro_error::{abort, proc_macro_error};
+use quote::quote;
+use syn::{parse_macro_input, ItemFn, Path, ReturnType, Signature, Type};
+
+/// Used to annotate user-defined function which starts the execution of smartcontract
+#[proc_macro_error]
+#[proc_macro_attribute]
+pub fn iroha_wasm(_: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        attrs,
+        vis,
+        sig,
+        block,
+    }: ItemFn = parse_macro_input!(item as ItemFn);
+
+    verify_function_signature(&sig);
+    let fn_name = &sig.ident;
+
+    quote! {
+        #[no_mangle]
+        unsafe extern "C" fn _iroha_wasm_main(ptr: u32, len: u32) {
```

Comment:
it is UB for two crates to export an item with the same name. When using `no_mangle` compiler cannot guarantee that this overlap will not occur. This usually means that it's not specified which of the items will be linked to. This is the reason why I prepended `_iroha_wasm` to all `no_mangle` function. `no_mangle` is a bit dangerous

https://github.com/rust-lang/rust/issues/28179

Reply:
Do you think `#[wasm_bindgen]` instead of `#[no_mangle]` bring something for this?
I [confirmed](https://github.com/s8sato/mangle/blob/main/src/main.rs) it allows to export items of the same name in the different modules:

FYI:

* https://docs.wasmtime.dev/wasm-rust.html?highlight=bindgen#exporting-rust-functionality

  > Rust functions must be #[no_mangle] to show up in the final binary, but if you're using #[wasm_bindgen] that will happen automatically for you.

  > Note that wasm-bindgen was originally developed for JS and usage in a browser, but a subset of its implementation (such as arguments which are strings) are supported for WebAssembly interface types.

* https://rustwasm.github.io/wasm-bindgen/contributing/design/describe.html?highlight=no_mangle#communicating-types-to-wasm-bindgen

Reply:
No. `#[wasm_bindgen]` requires `cargo wasi build --release` and this does not allows that

Reply:
can we close this? and I don't think this would solve the issue. They would just attach `no_mangle` to the function but the linking problem would persist.